lib/std/child_process.zig

```diff
@@ -1094,6 +1090,235 @@ pub const ChildProcess = struct {
+fn windowsCreateProcessPathExt(
+    allocator: mem.Allocator,
+    dir_buf: *std.ArrayListUnmanaged(u16),
+    app_buf: *std.ArrayListUnmanaged(u16),
+    pathext: [:0]const u16,
+    cmd_line: [*:0]u16,
+    envp_ptr: ?[*]u16,
+    cwd_ptr: ?[*:0]u16,
+    lpStartupInfo: *windows.STARTUPINFOW,
+    lpProcessInformation: *windows.PROCESS_INFORMATION,
+) !void {
+    const app_name_len = app_buf.items.len;
+    const dir_path_len = dir_buf.items.len;
+
+    if (app_name_len == 0) return error.FileNotFound;
+
+    defer app_buf.shrinkRetainingCapacity(app_name_len);
+    defer dir_buf.shrinkRetainingCapacity(dir_path_len);
+
+    // The name of the game here is to avoid CreateProcessW calls at all costs,
+    // and only ever try calling it when we have a real candidate for execution.
+    // Secondarily, we want to minimize the number of syscalls used when checking
+    // for each PATHEXT-appended version of the app name.
+    //
+    // An overview of the technique used:
+    // - Open the search directory for iteration (either cwd or a path from PATH)
+    // - Use NtQueryDirectoryFile with a wildcard filename of `<app name>*` to
+    //   check if anything that could possibly match either the unappended version
+    //   of the app name or any of the versions with a PATHEXT value appended exists.
+    // - If the wildcard NtQueryDirectoryFile call found nothing, we can exit early
+    //   without needing to use PATHEXT at all.
+    //
+    // This allows us to use a <open dir, NtQueryDirectoryFile, close dir> sequence
+    // for any directory that doesn't contain any possible matches, instead of having
+    // to use a separate look up for each individual filename combination (unappended +
+    // each PATHEXT appended). For directories where the wildcard *does* match something,
+    // we only need to do a maximum of <number of supported PATHEXT extensions> more
+    // NtQueryDirectoryFile calls.
+
+    var dir = dir: {
+        if (fs.path.isAbsoluteWindowsWTF16(dir_buf.items[0..dir_path_len])) {
+            const prefixed_path = try windows.wToPrefixedFileW(dir_buf.items[0..dir_path_len]);
+            break :dir fs.cwd().openDirW(prefixed_path.span().ptr, .{}, true) catch return error.FileNotFound;
+        }
+        // needs to be null-terminated
+        try dir_buf.append(allocator, 0);
+        defer dir_buf.shrinkRetainingCapacity(dir_buf.items[0..dir_path_len].len);
+        const dir_path_z = dir_buf.items[0 .. dir_buf.items.len - 1 :0];
+        break :dir std.fs.cwd().openDirW(dir_path_z.ptr, .{}, true) catch return error.FileNotFound;
+    };
+    defer dir.close();
+
+    // Add wildcard and null-terminator
+    try app_buf.append(allocator, '*');
+    try app_buf.append(allocator, 0);
+    const app_name_wildcard = app_buf.items[0 .. app_buf.items.len - 1 :0];
+
+    // Enough for the FILE_DIRECTORY_INFORMATION + (NAME_MAX UTF-16 code units [2 bytes each]).
```

Comment:
This doesnt explain why 2*NAME_MAX. context https://github.com/ziglang/zig/blob/master/lib/std/os/windows.zig#L3142